index.js

```diff
@@ -283,6 +294,126 @@ class Conf {
 		this.store = store;
 	}
 
+	_serializeExtraTypes(value) {
+		// Base cases
+		for (const type of this.extraTypes) {
+			if (type.isInstance(value)) {
+				return {
+					[TYPE_KEY]: type.name,
+					value: type.convertFrom(value)
+				};
+			}
+		}
+
+		// Value is a base type
+		if (typeof value !== 'object') {
+			return value;
+		}
+
+		// Sub-object handling of extra types
+		const stack = [];
+		const objectSeenList = [];
+
+		stack.push({current: value, parent: null, key: null});
+		let circularReferenceCounter = 1;
+		while (stack.length > 0) {
+			const {current, parent, key} = stack.pop();
+			if (objectSeenList.indexOf(current) === -1) {
+				objectSeenList.push(current);
+			} else {
+				// Current object is reference to an object already parsed
+				// current is the circular reference
+				// seenList.indexOf(current) is the object it's pointing to
+				const referenceTarget = objectSeenList[objectSeenList.indexOf(current)];
+				parent[key] = {
+					[TYPE_KEY]: 'circularObject',
+					value: referenceTarget[CIRCULAR_REF_KEY] || circularReferenceCounter
+				};
+				if (referenceTarget[CIRCULAR_REF_KEY] === undefined) {
+					referenceTarget[CIRCULAR_REF_KEY] = circularReferenceCounter;
+					circularReferenceCounter++;
+				} else {
+					console.log(referenceTarget[CIRCULAR_REF_KEY]);
+				}
+
+				console.log('skipping pontentially circulation');
+				continue;
+			}
+
+			for (const [currentItemKey, currentItemValue] of Object.entries(current)) {
+				if (typeof currentItemValue === 'object') {
+					let didConvert = false;
+					for (const type of this.extraTypes) {
+						if (type.isInstance(currentItemValue)) {
+							current[currentItemKey] = {
+								[TYPE_KEY]: type.name,
+								value: type.convertFrom(currentItemValue)
+							};
+							didConvert = true;
+							break;
+						}
+					}
+
+					if (!didConvert) {
+						// Check sub-objects
+						stack.push({current: current[currentItemKey], parent: current, key: currentItemKey});
+					}
+				}
+			}
+		}
+
+		return value;
+	}
```

Comment:
This method is pretty large. I would look into splitting it into multiple methods and maybe something here is generic enough to be moved out into a generic utility function. Same with `_deserializeExtraTypes`.